pkg/services/object/get/ec.go

```diff
@@ -0,0 +1,399 @@
+		//  Keep in sync with https://github.com/nspcc-dev/neofs-node/pull/3466.
+		payload := iec.ConcatDataParts(rule, hdr.PayloadSize(), parts)
+		hdr.SetPayload(payload)
+		return hdr, nil
+	}
+
+	eg, gCtx = errgroup.WithContext(ctx)
+	failCounter.Store(0)
+	var okCounter atomic.Uint32
+	for i := range rule.ParityPartNum {
+		partIdx := int(rule.DataPartNum + i)
+		eg.Go(func() error {
+			_, part, err := s.getECPart(gCtx, cnr, parent, sTok, bTok, rule, ruleIdx, sortedNodes, partIdx)
+			if err != nil {
+				if errors.Is(err, apistatus.ErrObjectAlreadyRemoved) || errors.Is(err, gCtx.Err()) {
+					return err
+				}
+				if failed := failCounter.Add(1); failed+uint32(rem) > uint32(rule.ParityPartNum) {
+					return tooManyPartsUnavailableError(failed)
+				}
+				s.log.Info("failed to get EC parity part",
+					zap.Stringer("container", cnr), zap.Stringer("object", parent), zap.Stringer("rule", rule),
+					zap.Int("ruleIdx", ruleIdx), zap.Int("partIdx", partIdx), zap.Error(err),
+				)
+				return nil
+			}
+
+			parts[partIdx] = part
+			if okCounter.Add(1) >= uint32(rem) {
+				return errInterrupt
+			}
+
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil && !errors.Is(err, errInterrupt) {
+		return object.Object{}, err
+	}
+
+	if rem = islices.CountNilsInTwoDimSlice(parts); rem > int(rule.ParityPartNum) {
+		return object.Object{}, tooManyPartsUnavailableError(rem)
+	}
+
+	payload, err := iec.Decode(rule, hdr.PayloadSize(), parts)
+	if err != nil {
+		return object.Object{}, fmt.Errorf("decode payload from parts: %w", err)
+	}
+
+	hdr.SetPayload(payload)
+	return hdr, nil
+}
+
+// reads object for EC part which should produced for parent object and
+// distributed sortedNodes according to EC rule with index = ruleIdx in cnr
+// policy. Returns parent object header and part payload. Payload slice is nil
+// iff parent payload is empty.
+//
+// Returns [apistatus.ErrObjectAlreadyRemoved] if the object was marked for
+// removal.
+//
+// Can return [context.Canceled] from the passed ctx only.
+func (s *Service) getECPart(ctx context.Context, cnr cid.ID, parent oid.ID, sTok *session.Object, bTok *bearer.Token,
+	rule iec.Rule, ruleIdx int, sortedNodes []netmap.NodeInfo, partIdx int) (object.Object, []byte, error) {
+	var partHdr object.Object
+	var rc io.ReadCloser
+
+	pi := iec.PartInfo{
+		RuleIndex: ruleIdx,
+		Index:     partIdx,
+	}
+
+	var err error
+	var local, localDone bool
+	for i := range iec.NodeSequenceForPart(partIdx, int(rule.DataPartNum+rule.ParityPartNum), len(sortedNodes)) {
+		local = !localDone && s.neoFSNet.IsLocalNodePublicKey(sortedNodes[i].PublicKey())
+		if local {
+			localDone = true
+		}
+
+		if local {
+			partHdr, rc, err = s.localObjects.GetECPart(cnr, parent, pi)
+		} else {
+			partHdr, rc, err = s.getECPartFromNode(ctx, cnr, parent, sTok, bTok, pi, sortedNodes[i])
+		}
+		if err == nil {
+			break
+		}
+		if errors.Is(err, apistatus.ErrObjectAlreadyRemoved) {
```

Comment:
this is a correct and expected operation's result, will we close `rc` in this case?